core/chains/evm/gas/block_history_estimator.go

```diff
@@ -304,76 +319,59 @@ func (b *BlockHistoryEstimator) BumpLegacyGas(_ context.Context, originalGasPric
 	return bumpedGasPrice, gasLimit, err
 }
 
-// checkConnectivity detects if the transaction is not being included due to
-// some kind of mempool propagation or connectivity issue rather than
-// insufficiently high pricing and returns error if so
-func (b *BlockHistoryEstimator) checkConnectivity(attempts []EvmPriorAttempt) error {
+// haltBumping prevents transactions from excessively bumping if an existing attempt's price is above a configurable percentile
+// This check is required in case the transaction is not being included due to some kind of mempool propagation
+// or connectivity issue rather than insufficiently high pricing
+func (b *BlockHistoryEstimator) haltBumping(attempts []EvmPriorAttempt) error {
 	percentile := int(b.bhConfig.CheckInclusionPercentile())
-	// how many blocks since broadcast?
-	latestBlockNum := b.getCurrentBlockNum()
-	if latestBlockNum == nil {
-		b.logger.Warn("Latest block is unknown; skipping inclusion check")
-		// can't determine anything if we don't have/know latest block num yet
-		return nil
+	// Get latest CheckInclusionPercentile prices to use in the checks below
+	var maxGasPrice *assets.Wei
+	var maxTipCap *assets.Wei
+	ok := b.IfStarted(func() {
+		maxGasPrice = b.getMaxPercentileGasPrice()
+		maxTipCap = b.getMaxPercentileTipCap()
+	})
+	if !ok {
+		return errors.New("BlockHistoryEstimator is not started; do not have max gas to allow bumping")
 	}
-	expectInclusionWithinBlocks := int(b.bhConfig.CheckInclusionBlocks())
-	blockHistory := b.getBlocks()
-	if len(blockHistory) < expectInclusionWithinBlocks {
-		b.logger.Warnf("Block history in memory with length %d is insufficient to determine whether transaction should have been included within the past %d blocks", len(blockHistory), b.bhConfig.CheckInclusionBlocks())
+	if !b.initialFetch.Load() {
+		return errors.New("BlockHistoryEstimator has not finished the first gas estimation yet, likely because a failure on start")
+	}
+	// Return nil to allow bumping to proceed if gas price is nil or if EIP1559 is enabled and tip cap is nil
+	// Do not halt bumping in case this is an issue with the head tracker to prevent transactions from getting backed up
+	if maxGasPrice == nil || (b.eConfig.EIP1559DynamicFees() && maxTipCap == nil) {
+		b.logger.Debugf("%d percentile price is not set. This is likely because there aren't any valid transactions to estimate from. Allowing bumping to proceed.", percentile)
 		return nil
 	}
+	// Get the latest CheckInclusionBlocks from block history for fee cap check below
+	blockHistory := b.getBlocks()
+	blockRange := mathutil.Min(len(blockHistory), int(b.bhConfig.CheckInclusionBlocks()))
+	startIdx := len(blockHistory) - blockRange
+	checkInclusionBlocks := blockHistory[startIdx:]
+	// Check each attempt for any with a gas price or tip cap (if EIP1559 type) exceeds the latest CheckInclusionPercentile prices
 	for _, attempt := range attempts {
 		if attempt.BroadcastBeforeBlockNum == nil {
```

Comment:
I left this check in here as part of a belt and braces approach since it's still an assumption violation. Since we don't use `attempt.BroadcastBeforeBlockNum` anymore, you're right that this check isn't really protecting anything except halting bumping if a tx has an attempt without `BroadcastBeforeBlockNum` set.

Reply:
Actually I had a major suspect this code is causing problems.

Inside our confirmer, we are setting BroadcastBeforeBlockNum to null here:
https://github.com/smartcontractkit/chainlink/blob/d675d864f0e6f33c783bfed17fe31b2c127eb51d/common/txmgr/confirmer.go#L857

I couldn't find where we do again set the value back to a valid block.
Could you check that, and ensure it isn't set to null forever?

Reply:
Otherwise, I'm in favor of removing this check

Reply:
I believe after this code the Confirmer tries broadcast the transaction. Whether it's successful or fails, it saves the attempt without a `BroadcastBeforeBlockNum`. Then on the next call of `processHead`, we set the block num for all `broadcast` attempts here:
https://github.com/smartcontractkit/chainlink/blob/58d73ecf618ac39c37f767e70c4e6d6a51eaba59/common/txmgr/confirmer.go#L288-L295 The only time an attempt wouldn't have a `BroadcastBeforeBlockNum` is if it is either `in_progress` or `insufficient_funds` state. But txs with an `insufficient_funds` attempt are never passed to bumping and attempts aren't marked as `in_progress` anywhere in the call stack before bumping.

If we want to ensure we don't bump for txs with either `in_progress` or `insufficient_funds` attempts, I can change this validation to check that all attempts are `broadcast` state. Seems like it would be functionally the same as the  `BroadcastBeforeBlockNum` check but much more straight forward. I'm also ok removing this check but seems like a good check to have for a belts and braces approach since we would want to block bumping if a tx has a non-broadcast state attempt.

Reply:
Thanks for the validation. Looks good.